### PR TITLE
Basic support for multipart/form-data for node client

### PR DESCRIPTION
**Pull request by vincentjames501**
For #98. Looks to be working fine at least for the node client. I didn't have time to write proper tests, but this is at least a mockup of supporting multipart in node.

### CODE REVIEW

File: client/node.js [hunk 1]
```diff
@@ -84,10 +84,17 @@
...
 				if (!headers['Content-Length']) {
-					headers['Content-Length'] = entity ? Buffer.byteLength(entity, 'utf8') : 0;
+					if (isMultiPart) {
+						headers['Content-Length'] = entity ? entity.getLengthSync() : 0;
+					} else {
+						headers['Content-Length'] = entity ? Buffer.byteLength(entity, 'utf8') : 0;
+					}
+				}
+				if (isMultiPart) {
+					headers['Content-Type'] = entity.getHeaders()['content-type'];
```
Comment (vincentjames501) [inline client/node.js]: Just noticed, this needs to make sure entity is not null.

**Comment by scothis** [comment]
What is the type of entity in this case? I see the check for the request to be multipart based on the content-type, but never a check for the entity to be of a particular type.

**Comment by scothis** [comment]
As, I see what you're trying to do. There are a few issues, which you can see as CI fell over.

1. the client can not assume the mime interceptor was applied to transform the entity into a given type. The client should be able to handle whatever is thrown at it, gracefully. As a goal for rest.js is to keep the clients as minimal as possible, in many cases, this means that the entity needs to be transformed into something friendly before it arrives at the root client. Removing the logic from the root client also makes the behavior portable to other clients, avoiding platform specific features.
2. At this time, I'm not willing to accept a new required dependency for rest.js. In particular a dependency which is specific to a single platform. As rest.js also needs to work in the browser, without a compile step, depending on a node library directly is a no-go.

I think you can be successful with what you're trying to achieve without modifying the node client. You can use an interceptor to transform the entity and headers. After all a multipart http request is just an http request, with a specially encoded entity. This will be easier with 2.0, but should be doable now.

I'd be happy to review your progress.

**Comment by vincentjames501** [comment]
I should've specified that this was just intended for a discussion, not to actually be merged as this was clearly a hack job. 

There are a couple of issues with the current node client then. The client makes the assumption that the `entity` is a String (if its defined) as ```Buffer.byteLength(entity, 'utf8')``` and ```clientRequest.write(entity);``` with blow up with other types. So this should also be moved to it's own interceptor in my opinion.

I'm also not sure if I understand the hesitation of adding another required dependency for node. Adding it to just the ```package.json``` and adding the proper logic should make it a safe add and simplify client interaction for doing multipart upload.

**Comment by scothis** [comment]
No worries. Let's continue the conversation

If you base64 encode the binary portions of the request, getting that bit of the node client should work. This is one of the points that will be easier with rest.js 2.0 once we add proper support for binary entities.

Adding a dependency to package.json doesn't make that dependency node specific. For example, jspm will read that dependency metadata for rest.js even though the result can be consumed in a browser. Keeping the dependency graph small makes it easier for browser consumers who are using require.js or curl.js and need to configure module paths. It's admittedly less of a concern with newer loaders like System.js via jspm or RaveJS. As more and more devs move to tools that simplify complex loading patterns or introduce a build step, I'm less concerned, but we're not their yet. For 2.0, I'm considering authoring in ES6, which would require a transpile for current environments.

**Pull request closed without merging**